packages/ngtools/webpack/src/ivy/plugin.ts

```diff
@@ -553,73 +553,81 @@ export class AngularWebpackPlugin {
-      const OptimizeFor =
-        // eslint-disable-next-line @typescript-eslint/no-explicit-any
-        (this.compilerCli as any).OptimizeFor ??
-        require('@angular/compiler-cli/src/ngtsc/typecheck/api').OptimizeFor;
-
-      // Collect new Angular diagnostics for files affected by changes
-      const optimizeDiagnosticsFor =
-        affectedFiles.size <= DIAGNOSTICS_AFFECTED_THRESHOLD
-          ? OptimizeFor.SingleFile
-          : OptimizeFor.WholeProgram;
-      for (const affectedFile of affectedFiles) {
-        const angularDiagnostics = angularCompiler.getDiagnosticsForFile(
-          affectedFile,
-          optimizeDiagnosticsFor,
+        // Temporary workaround during transition to ESM-only @angular/compiler-cli
+        // TODO_ESM: This workaround should be removed prior to the final release of v13
+        //       and replaced with only `this.compilerCli.OptimizeFor`.
+        const OptimizeFor =
+          // eslint-disable-next-line @typescript-eslint/no-explicit-any
+          (this.compilerCli as any).OptimizeFor ??
+          require('@angular/compiler-cli/src/ngtsc/typecheck/api').OptimizeFor;
+
+        // Collect new Angular diagnostics for files affected by changes
+        const optimizeDiagnosticsFor =
+          affectedFiles.size <= DIAGNOSTICS_AFFECTED_THRESHOLD
+            ? OptimizeFor.SingleFile
+            : OptimizeFor.WholeProgram;
+        for (const affectedFile of affectedFiles) {
+          const angularDiagnostics = angularCompiler.getDiagnosticsForFile(
+            affectedFile,
+            optimizeDiagnosticsFor,
+          );
+          diagnosticsReporter(angularDiagnostics);
+          this.sourceFileCache?.updateAngularDiagnostics(affectedFile, angularDiagnostics);
+        }
+
+        return this.createFileEmitter(
+          builder,
+          mergeTransformers(angularCompiler.prepareEmit().transformers, transformers),
+          getDependencies,
+          (sourceFile) => {
+            this.requiredFilesToEmit.delete(normalizePath(sourceFile.fileName));
+            angularCompiler.incrementalDriver.recordSuccessfulEmit(sourceFile);
+          },
         );
-        diagnosticsReporter(angularDiagnostics);
-        this.sourceFileCache?.updateAngularDiagnostics(affectedFile, angularDiagnostics);
-      }
+      })
+      .catch((err) => (err instanceof Error ? err : new Error(err)));
 
-      return this.createFileEmitter(
-        builder,
-        mergeTransformers(angularCompiler.prepareEmit().transformers, transformers),
-        getDependencies,
-        (sourceFile) => {
-          this.requiredFilesToEmit.delete(normalizePath(sourceFile.fileName));
-          angularCompiler.incrementalDriver.recordSuccessfulEmit(sourceFile);
-        },
-      );
-    });
     const analyzingFileEmitter: FileEmitter = async (file) => {
       const innerFileEmitter = await pendingAnalysis;
 
+      if (innerFileEmitter instanceof Error) {
```

Comment:
It's a little weird that `pendingAnalysis` is of type `Promise<FileEmitter | Error>`, especially since it could still throw an error and `Promise` already has semantics for how that works.

I'm wondering if this would be cleaner to either actually `throw` the error (maybe as some subtype we can use an `instanceof` on) or return a proper discriminated union and `switch` on the two possible values? I think that would be much more clear about the actual contract of this promise and be easier to follow.